proposals/p2173.md

```diff
@@ -0,0 +1,489 @@
+    `.A` is named as a member of a type with this constraint, it is rewritten to
+    `T`. This means that the interface of the type changes, and the behavior is
+    in all respects as if `T` had been specified directly.
+-   `where X == Y` specifies a _same-type_ constraint. This is treated much like
+    `where X is SameAs(Y)`. Here, `SameAs` is a hypothetical constraint that is
+    reflexive (`T is SameAs(T)` for all `T`), commutative (`T is SameAs(U)` if
+    and only if `U is SameAs(T)`), and not implementable by the developer.
+    Moreover, if `F` is any pure type function, `T is SameAs(U)` implies
+    `F(T) is SameAs(F(U)`). `observe` statements can be used to form transitive
+    `SameAs` relations. Same-type constraints are not used automatically by the
+    language rules for any purpose, but there is an `ImplicitAs(T)` impl for
+    types implementing `SameAs(T)`.
+
+A type implements `C where .A = T` if and only if it implements
+`C where .A == T` – the inhabitants of these two type-of-types are the same, but
+they are different type-of-types and provide different interfaces for
+corresponding type values.
+
+## Details
+
+### Rewrite constraints
+
+In a rewrite constraint, the left-hand operand of `=` must be the dotted name of
+an associated constant. This notation is permitted anywhere a constraint can be
+written, and results in a new constraint with a different interface: the named
+member effectively no longer names an associated constant of the constrained
+type, and is instead treated as a rewrite rule the expands to the right-hand
+side of the constraint, with any mentioned parameters substituted into that
+type.
+
+```
+interface Container {
+  let Element:! Type;
+  let Slice:! Container where .Element = Element;
+}
+// `T.Slice.Element` rewritten to `T.Element` because type of `T.Slice` says `.Element = Element`.
+// `T.Element` rewritten to `i32` because type of `T` says `.Element = i32`.
+fn A[T:! Container where .Element = i32](x: T.Slice.Element) {}
+```
+
+Rewrites aren't performed on the left-hand side of such an `=`, so
+`where .A = .B and .A = C` is not rewritten to `where .A = .B and .B = C`.
+Instead, such a `where` clause is invalid unless each rule for `.A` specifies
+the same rewrite.
+
+When an `=` appears within an interface, it is possible to form a cycle where a
+type becomes more complex with each iteration. In order to avoid this problem,
+if, when substituting into the right-hand side of such an `=`, the same rewrite
+is performed again, the program is invalid.
+
+```
+interface Container {
+  let Element:! Type;
+  // (1), bad constraint
+  let Slice:! Container where .Element = .Slice.Element;
+}
+// Error, `T.Slice.Element` rewritten to
+// `T.Slice.Slice.Element` by (1), then to
+// `T.Slice.Slice.Slice.Element` by (1)
+fn A[T:! Container](x: T.Slice.Element) {}
+```
+
+It is valid for the same rewrite rule to be used multiple times within a type
```

Comment:
````suggestion
Here, "same rewrite" is determined by whether the same associated constant in
the same interface member is rewritten, ignoring any parameterization of the
interface.

```
interface ContainerOf(T:! Type) {
  let Iterator:! Type;
  // (2), bad constraint
  let Pointers:! ContainerOf(T*) where .Iterator = .Pointers.Iterator;
}
// Error, `T.Pointers.Iterator` rewritten to
// `T.Pointers.Pointers.Iterator` by (2) in `ContainerOf(i32)`, then to
// `T.Pointers.Pointers.Pointers.Iterator` by (2) in `ContainerOf(i32*)`
fn B[T:! ContainerOf(i32)](x: T.Pointers.Iterator) {}
```

It is valid for the same rewrite rule to be used multiple times within a type
````

Reply:
New idea: use the location of the `=` and also the location of the `:!` if any.